### ISSUE TITLE
Chain Sync Client: maintain "forks within k" invariant

### ISSUE THREAD

**Issue by edsko** [input-output-hk/ouroboros-network#579]
The chain sync client must ensure that the chain fragment we are maintaining for the upstream peer _must_ always satisfy the invariant that

```haskell
    -- INVARIANT: 
    --
    -- >           our tip
    -- >             v
    -- >   /--* .... *
    -- >   |
    -- > --*
    -- >   |
    -- >   \--* .... *
    -- >        fragment tip
    --
    -- The distance from our tip to the intersection between our chain and
    -- the fragment maintained for the upstream node cannot exceed @k@ blocks.
    -- When this intervariant cannot be maintained, the upstream node is on
    -- a fork that is too distant and we should disconnect.
```

It currently does not do this.

**Comment by dcoutts** [comment]
We have concluded the actual requirement is somewhat weaker. The block fetch can tolerate candidates that do not intersect within K. The requirement is about placing a bound on how long we will continue to spend resources on following a peers chain if we conclude that it is not interesting to us.

**Comment by mrBliss** [comment]
Addressed by #1081.
